<img src="https://gallery.mailchimp.com/f98d5ac0a3fbbdcdda35136ab/images/370b1236-88aa-4969-b768-cd07bd006d8c.png">

# Computer Vision APIs (with [Clarifai](https://clarifai.com/))

<img src='../assets/clarifai-logo.png' style='width:300px'>

First, let us install the relevant package for us to interact with Clarifai's API.

In [ ]:
!pip install clarifai --upgrade

Now that we have installed the package, we call upon the relevant package to to initialise the API. Replace `YOUR_API_KEY` with the API keys obtained from the [Clarifai apps portal](https://clarifai.com/apps). What this means is that you have to first create an app through the portal.

In [1]:
API_KEY = 'YOUR_CLARIFAI_API_KEY_HERE'

Now that you have stored your API key in a variable, let us initialise the application.

In [2]:
from clarifai.rest import ClarifaiApp

app = ClarifaiApp(api_key=API_KEY)

From the initialised app, we have to define which model we would like to utilise. Clarifai has a selection of [pretrained models](https://clarifai.com/models) that we can utilise. For now, let us stick with the [general model](https://clarifai.com/models/general-image-recognition-model-aaa03c23b3724a16a56b629203edc62c) which allows us to do generic computer vision detection of subjects.

In [3]:
model = app.public_models.general_model

Now, we have the model provide us with predictions pertaining to the image link that we have provided. 

The image link brings us to this image:

<img src='https://images.pexels.com/photos/326012/pexels-photo-326012.jpeg' style='width:300px'>

In [4]:
# Store the response from the API into a variable
response = model.predict_by_url(url='https://images.pexels.com/photos/326012/pexels-photo-326012.jpeg')

Let us observe the response provided by the API.

In [5]:
response

{'status': {'code': 10000,
  'description': 'Ok',
  'req_id': '3049f6256b5342d5843aae8a355b682d'},
 'outputs': [{'id': '9fbb81ff24774453abfe526c378c23a8',
   'status': {'code': 10000, 'description': 'Ok'},
   'created_at': '2019-05-12T22:07:08.607761322Z',
   'model': {'id': 'aaa03c23b3724a16a56b629203edc62c',
    'name': 'general',
    'created_at': '2016-03-09T17:11:39.608845Z',
    'app_id': 'main',
    'output_info': {'message': 'Show output_info with: GET /models/{model_id}/output_info',
     'type': 'concept',
     'type_ext': 'concept'},
    'model_version': {'id': 'aa9ca48295b37401f8af92ad1af0d91d',
     'created_at': '2016-07-13T01:19:12.147644Z',
     'status': {'code': 21100, 'description': 'Model trained successfully'}},
    'display_name': 'General'},
   'input': {'id': '037fc05e28f145b284434979d831ca3d',
    'data': {'image': {'url': 'https://images.pexels.com/photos/326012/pexels-photo-326012.jpeg'}}},
   'data': {'concepts': [{'id': 'ai_l6FJp7Lf',
      'name': 'rabbit'

The output above can be convoluted at first look but let us filter the information that we need.

In [6]:
response['outputs'][0]['data']['concepts']

[{'id': 'ai_l6FJp7Lf',
  'name': 'rabbit',
  'value': 0.9987200498580933,
  'app_id': 'main'},
 {'id': 'ai_l2mHxdfZ',
  'name': 'bunny',
  'value': 0.9976458549499512,
  'app_id': 'main'},
 {'id': 'ai_K5cG7T4B',
  'name': 'hare',
  'value': 0.9890360236167908,
  'app_id': 'main'},
 {'id': 'ai_4CRlSvbV',
  'name': 'cute',
  'value': 0.9887425899505615,
  'app_id': 'main'},
 {'id': 'ai_N6BnC4br',
  'name': 'mammal',
  'value': 0.9868383407592773,
  'app_id': 'main'},
 {'id': 'ai_mp9KG5QH',
  'name': 'grass',
  'value': 0.9835637807846069,
  'app_id': 'main'},
 {'id': 'ai_Cj7dpk3Z',
  'name': 'Easter',
  'value': 0.9818881154060364,
  'app_id': 'main'},
 {'id': 'ai_SzsXMB1w',
  'name': 'animal',
  'value': 0.9669061899185181,
  'app_id': 'main'},
 {'id': 'ai_wqgGDjK4',
  'name': 'fur',
  'value': 0.9618996381759644,
  'app_id': 'main'},
 {'id': 'ai_HsV5wmGX',
  'name': 'rodent',
  'value': 0.9500438570976257,
  'app_id': 'main'},
 {'id': 'ai_gPNNX7LT',
  'name': 'pet',
  'value': 0.942628

We can see that the model gives us several possible descriptions of the image that was provided to it and for each of them, a value is given. This value pertains to its confidence level in regards to the descriptions. For example, the prediction 'rabbit' has the value '0.9987' so the model is 99.87% confident that the image provided is of a rabbit.

Below, we will try to convert the output into a `pandas` dataframe format which is more readable.

In [7]:
from pandas.io.json import json_normalize
import pandas as pd

In [8]:
clarif_res_df = json_normalize(response['outputs'][0]['data']['concepts'])
clarif_res_df

app_id           id       name     value
0    main  ai_l6FJp7Lf     rabbit  0.998720
1    main  ai_l2mHxdfZ      bunny  0.997646
2    main  ai_K5cG7T4B       hare  0.989036
3    main  ai_4CRlSvbV       cute  0.988743
4    main  ai_N6BnC4br     mammal  0.986838
5    main  ai_mp9KG5QH      grass  0.983564
6    main  ai_Cj7dpk3Z     Easter  0.981888
7    main  ai_SzsXMB1w     animal  0.966906
8    main  ai_wqgGDjK4        fur  0.961900
9    main  ai_HsV5wmGX     rodent  0.950044
10   main  ai_gPNNX7LT        pet  0.942629
11   main  ai_786Zr311  no person  0.939948
12   main  ai_R67cVW5t        sit  0.926562
13   main  ai_BpnQsTtg      young  0.926509
14   main  ai_xWCNmTbf   adorable  0.911700
15   main  ai_tBcWlsCp     nature  0.910618
16   main  ai_HkqZwBbr     little  0.902726
17   main  ai_cNfpjTnr      furry  0.888695
18   main  ai_fbqvMwRm   wildlife  0.883059
19   main  ai_QGHBfsBT      funny  0.875218

## Creating a model

The example that we have run above relies on pretrained classifiers that Clarifai has set up and train beforehand. What we can also do with Clarifai is to create our custom classifiers, while tapping on to the deep learning architectures and model that Clarifai has. One can keep in mind that this is the usage of transfer learning as we are going to utilise an already existing architecture and simply tweaking it to our own use case.

In this example that we are about to carry out, we are going to create a custom model that allows us to predict whether a certain image shows the following outputs:
+ rabbit
+ not rabbit

<img src='https://images.pexels.com/photos/326012/pexels-photo-326012.jpeg' style='width:300px'>
<img src='https://images.pexels.com/photos/151511/pexels-photo-151511.jpeg' style='width:300px'>

In order to train our custom model, we first have to upload some images to train the model on.

In [9]:
# Upload an image that pertains to 'rabbit' concept/classification
app.inputs.create_image_from_url(url='https://images.pexels.com/photos/326012/pexels-photo-326012.jpeg', concepts=['rabbit'])
# Upload an image that pertains to 'not rabbit' concept/classification
app.inputs.create_image_from_url(url='https://images.pexels.com/photos/151511/pexels-photo-151511.jpeg', not_concepts=['rabbit'])

After uploading the images, we initialise a custom model and name it. To this model, we can associate the concepts/classifications that we are tasking for it. In this case, we just associate it with the concept `rabbit`.

In [10]:
custom_model = app.models.create(model_id="rabbit_mod", concepts=["rabbit"])

We run the following to start the training process; the model will train on the images that you have uploaded above.

In [11]:
custom_model.train()

These custom models once trained, can be stored and readied for whenever we intend to use them. Below is how we call upon them.

In [12]:
imported_custom_model = app.models.get('rabbit_mod')

Here is how you use the model to conduct a prediction on an image by URL.

<img src='https://images.pexels.com/photos/53966/rabbit-palm-hand-snatch-53966.jpeg' style='width:500px'>

In [13]:
pred_res = imported_custom_model.predict_by_url('https://images.pexels.com/photos/53966/rabbit-palm-hand-snatch-53966.jpeg')

Let us examine the results.

In [14]:
pred_res['outputs'][0]['data']

{'concepts': [{'id': 'rabbit',
   'name': 'rabbit',
   'value': 0.0006965667707845569,
   'app_id': 'fc264822b6cb4e86ab62827edc3673fa'}]}

The confidence value for this image in whether it's a rabbit or not is very low even though the image clearly shows of a rabbit. This is predictable as we have trained the model using only two images when one should utilise way more.

For fun, examine the model's performance on the same image that it was trained on.

<img src='https://images.pexels.com/photos/326012/pexels-photo-326012.jpeg' style='width:300px'>

In [15]:
pred_res2 = imported_custom_model.predict_by_url('https://images.pexels.com/photos/326012/pexels-photo-326012.jpeg')

In [16]:
pred_res2['outputs'][0]['data']

{'concepts': [{'id': 'rabbit',
   'name': 'rabbit',
   'value': 0.9994441866874695,
   'app_id': 'fc264822b6cb4e86ab62827edc3673fa'}]}

The value is high is expected.

<center> <h1>~ End of section ~</h1> </center>